src/IO/H5/VolumeData.cpp

```diff
@@ -154,6 +162,308 @@ void append_element_extents_and_connectivity(
+    }
+    // Connectivity that already exists
+    expected_connectivity_length += element_connectivity_length * 8;
+    // Used for reserving the length of block_logical_coords
+    expected_number_of_grid_points += element_grid_points;
+  }
+
+  std::string grid_name_string = block_grid_names[0];
+  std::array<int, SpatialDim> h_ref_array = {};
+  size_t h_ref_previous_start_position = 0;
+  size_t additional_connectivity_length = 1;
+  for (size_t i = 0; i < SpatialDim; ++i) {
+    size_t h_ref_start_position =
+        grid_name_string.find('L', h_ref_previous_start_position + 1);
+    size_t h_ref_end_position =
+        grid_name_string.find('I', h_ref_start_position);
+    int h_ref = std::stoi(
+        grid_name_string.substr(h_ref_start_position + 1,
+                                h_ref_end_position - h_ref_start_position - 1));
+    h_ref_array[i] = h_ref;
+    additional_connectivity_length *= pow(2, h_ref + 1) - 1;
+    h_ref_previous_start_position = h_ref_start_position;
+  }
+
+  expected_connectivity_length +=
+      (additional_connectivity_length - block_extents.size()) * 8;
+
+  return std::tuple{expected_connectivity_length,
+                    expected_number_of_grid_points, h_ref_array};
+}
+
+// Generates the mesh
+template <size_t SpatialDim>
+Mesh<SpatialDim> generate_element_mesh(
+    const std::vector<std::string>& element_bases,
+    const std::vector<std::string>& element_quadratures,
+    const std::vector<size_t>& element_extents) {
+  std::array<Spectral::Basis, SpatialDim> basis_array = {};
+
+  std::array<Spectral::Quadrature, SpatialDim> quadrature_array = {};
+
+  std::array<size_t, SpatialDim> extents_array = {};
+
+  for (size_t i = 0; i < SpatialDim; i++) {
+    basis_array[i] = Spectral::to_basis(element_bases[i]);
+    quadrature_array[i] = Spectral::to_quadrature(element_quadratures[i]);
+    extents_array[i] = element_extents[i];
+  }
+
+  return Mesh<SpatialDim>{extents_array, basis_array, quadrature_array};
+}
+
+template <size_t SpatialDim>
+std::vector<std::array<double, SpatialDim>> generate_block_logical_coordinates(
+    const std::vector<std::array<double, SpatialDim>>&
+        element_logical_coordinates,
+    const std::string& grid_name,
+    const std::array<int, SpatialDim>& h_refinement_array) {
+  size_t grid_points_x_start_position = 0;
+  std::vector<std::array<double, SpatialDim>> block_logical_coordinates;
+  block_logical_coordinates.reserve(element_logical_coordinates.size());
+  std::vector<double> number_of_elements_each_direction;
+  number_of_elements_each_direction.reserve(SpatialDim);
+  std::vector<double> shift_each_direction;
+  shift_each_direction.reserve(SpatialDim);
+
+  for (size_t i = 0; i < SpatialDim; ++i) {
+    double number_of_elements = pow(2, h_refinement_array[i]);
+    number_of_elements_each_direction.push_back(number_of_elements);
+    size_t grid_points_start_position =
+        grid_name.find('I', grid_points_x_start_position + 1);
```

Comment:
Same question about number of grid points. Is there an issue here if you have one versus two digits? Same with refinement level and segment

Reply:
I believe this works the same as what I described for the parsing of the block number. Instead, we just use the fact that in a given string `[B#,I#,I#,I#]`, we abuse the fact that the number follows after the `I`, but ends before the `,`.

Reply:
With https://github.com/sxs-collaboration/spectre/pull/4380 I think you don't need this code here anymore, you can just use that, right?